runtime/src/bank.rs

```diff
@@ -897,77 +897,116 @@ impl Bank {
             let inflation = self.inflation.read().unwrap();
 
             (*inflation).validator(year) * self.capitalization() as f64 * period
-        };
+        } as u64;
+
+        let vote_balance_and_staked = self.stakes.read().unwrap().vote_balance_and_staked();
 
-        let validator_points = self.stakes.write().unwrap().claim_points();
-        let validator_point_value =
-            self.check_point_value(validator_rewards / validator_points as f64);
+        let validator_point_value = self.pay_validator_rewards(validator_rewards);
+
+        // this sysvar could be retired...
         self.update_sysvar_account(&sysvar::rewards::id(), |account| {
             sysvar::rewards::create_account(
                 self.inherit_sysvar_account_balance(account),
                 validator_point_value,
             )
         });
 
-        let validator_rewards = self.pay_validator_rewards(validator_point_value);
+        let validator_rewards_paid =
+            self.stakes.read().unwrap().vote_balance_and_staked() - vote_balance_and_staked;
+
+        // verify that we didn't pay any more than we expected to
+        assert!(validator_rewards >= validator_rewards_paid);
 
         self.capitalization
-            .fetch_add(validator_rewards as u64, Ordering::Relaxed);
+            .fetch_add(validator_rewards_paid, Ordering::Relaxed);
     }
 
-    /// iterate over all stakes, redeem vote credits for each stake we can
-    ///   successfully load and parse, return total payout
-    fn pay_validator_rewards(&mut self, point_value: f64) -> u64 {
-        let stake_history = self.stakes.read().unwrap().history().clone();
-        let mut validator_rewards = HashMap::new();
-
-        let total_validator_rewards = self
+    /// map stake delegations into resolved (pubkey, account) pairs
+    ///  returns a vector (has to be copied) of loaded
+    ///   ( (staker info) (voter info) )
+    ///
+    fn stake_delegation_accounts(&self) -> Vec<((Pubkey, Account), (Pubkey, Account))> {
```

Comment:
I love this helper fn. Reasoning got a lot easier. :)

Reply:
@ryoqun I think this approach still has merit, if the `store_account()` phase is moved until after all the stakes have been processed, or if the vote_account is modified while in the vector